### PR TITLE
[balsa] Add default-false runtime flag for BalsaParser.

### PR DESCRIPTION
**Pull request by bencebeky**
Also parametrize Http1CodecTests to meet required code coverage for balsa_parser.{h,cc}.
As a result, the following command runs all Http1CodecTests with both http-parser and BalsaParser:
   bazel test //test/common/http/http1:codec_impl_test
(Some BalsaParser tests are disabled for now.)

The runtime flag also allows developers to run tests other than Http1CodecTests locally
with BalsaParser enabled, for example:

bazel test //test/...:all \
  --test_arg=--runtime-feature-override-for-tests=envoy.reloadable_features.http1_use_balsa_parser

Tracking issue: #21245

Signed-off-by: Bence Béky <bnc@google.com>

Commit Message: Add default-false runtime flag for BalsaParser.
Additional Description:
Risk Level: low
Testing: //test/common/http/http1:codec_impl_test, more tests to come before flag is enabled
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

**Comment by bencebeky** [comment]
@alyssawilk PTAL

File: test/common/http/http1/codec_impl_test.cc [hunk 1]
```diff
@@ -894,15 +921,23 @@ TEST_F(Http1ServerConnectionImplTest, Http11AbsoluteEnabledNoOp) {
   expectHeadersTest(Protocol::Http11, true, buffer, expected_headers);
 }
 
-TEST_F(Http1ServerConnectionImplTest, Http11InvalidRequest) {
+TEST_P(Http1ServerConnectionImplTest, Http11InvalidRequest) {
+  if (parser_impl_ == ParserImpl::BalsaParser) {
```
Comment (alyssawilk) [inline test/common/http/http1/codec_impl_test.cc:925]: Is there a plan to resolve behavioral differences?
For example for HTTP/3 we had a macro which did this and a TODO at the macro definition to remove it after resolving all of the behavioral differences.

**Review by alyssawilk — COMMENTED** [review]
This looks fantastic!  Just the one test/TODO nit and you're good to go :-)

Please also fill out the commit fields in the PR description* as we use them for the final merge

*Testing:
Docs Changes:
Release Notes: etc

**Comment by alyssawilk** [comment]
Snow can you give a non-googler stamp? No need to do a full review of the test, just a conceptual LG

File: test/common/http/http1/codec_impl_test.cc (hunk 1 above)
Comment (bencebeky) [reply test/common/http/http1/codec_impl_test.cc:925]: Yes, the plan is to remove all differences.  I'm adding a TODO to each of the disabled tests for clarity.

**Review by alyssawilk — APPROVED** [review]
I think it's mildly easier to have the one TODO with a macro so it's really obvious if you can remove the macro or not but that's 100% a style thing so LGTM :-)

**Comment by alyssawilk** [comment]
CI failure looks real:
Parsers/Http1ServerConnectionImplTest.LargeTrailersRejectedIgnored/BalsaParser, where GetParam() = 4-byte object <01-00 00-00>
[  FAILED  ] Parsers/Http1ServerConnectionImplTest.LargeTrailerFieldRejectedIgnored/BalsaParser, 
/wait

**Review by snowp — DISMISSED** [review]
LGTM from a high level

**Comment by bencebeky** [comment]
> CI failure looks real: Parsers/Http1ServerConnectionImplTest.LargeTrailersRejectedIgnored/BalsaParser, where GetParam() = 4-byte object <01-00 00-00> [ FAILED ] Parsers/Http1ServerConnectionImplTest.LargeTrailerFieldRejectedIgnored/BalsaParser, /wait

Oops.  That's the result of #22240 that landed in the meanwhile.  I'm disabling those two tests as well for BalsaParser for the time being.  This will allow the current PR to land, and these BalsaParser issues will be addressed in later PRs.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**